Use configurable remote name for git commands

ilya-nurullin opened the pull request:
Closes #19403, and makes it possible to use any remote name in code snippets for an empty repository and pull request.
This change is very helpful to me, because I always use different name for my gitea remote.

Uses setting config module to store the value. Default is `origin` for backward compatibility.

### Screenshots
<details>
<summary>Empty repo</summary>
<img width="791" height="398" alt="image" src="https://github.com/user-attachments/assets/7214053d-a8dd-4e77-8c9d-78936d9859e0" />
</details>

<details>
<summary>Pull Request</summary>
<img width="591" height="452" alt="image" src="https://github.com/user-attachments/assets/ebc3d25c-5d6d-481d-819d-9706af3c5594" />
</details>

<details>
<summary>Settings page</summary>
<img width="1438" height="839" alt="image" src="https://github.com/user-attachments/assets/d92bfa2c-7adc-4efe-95fa-0c55ad13b3f5" />
</details>

hiifong commented:
This remote should be an optional configuration item with a default value of origin.

lunny commented:
Maybe it should be a per repository setting but not a global one.

ilya-nurullin commented:
> Does the title “Repository Snippets” in the configuration accurately reflect its purpose? It might be more appropriate under the Pull Request Settings section.

This setting affects two pages at once:
1. Empty repo page - I believe this is the most frequent use case for the setting
2. Pull Request page - you can do a merge using the button on the page, so snippets are hidden by default

Therefore, I would choose a more general name. I am open to any suggestions.

wxiaoguang commented on templates/admin/config_settings.tmpl:
```diff
@@ -39,4 +39,19 @@
+	<form class="ui form form-fetch-action" method="post" action="{{AppSubUrl}}/-/admin/config?key={{.SystemConfig.Repository.SnippetRemoteName.DynKey}}">
+		<div class="field" data-field-patched="true">
+			<label for="remote_name">{{ctx.Locale.Tr "admin.config.repository_snippets.remote_name"}}</label>
+			<input id="remote_name" name="value" value="{{.SystemConfig.Repository.SnippetRemoteName.Value ctx}}" maxlength="100" dir="auto" placeholder="{{.SystemConfig.Repository.SnippetRemoteName.Def}}">
```
I don't think the `Def` here is useful.

The value is always filled by `Value` when a user opens this page.

So I think we should remove the `Def` function.

ilya-nurullin replied:
The main reason for `Def` is being able to get the default value when user removes value on settings page. 
It is used here: https://github.com/go-gitea/gitea/pull/35172/files#diff-b3b66aef8cc83304cd5cfc661233e6c1c1be656fbd6cd1e09b114348f224ff32R251

In current architecture `func (value *Value[T]) Value(ctx context.Context) (v T)` function uses a default value only if it is not present in the database. So `null` in db, `nil` in go values are valid ones, and therefore the default value won't be used, and as a result there won't be any remote name on the page in snippets.

wxiaoguang replied:
> [](https://github.com/go-gitea/gitea/pull/35172/files#diff-b3b66aef8cc83304cd5cfc661233e6c1c1be656fbd6cd1e09b114348f224ff32R251)
> 
> In current architecture `func (value *Value[T]) Value(ctx context.Context) (v T)` function uses a default value only if it is not present in the database. So `null` in db, `nil` in go values are valid ones, and therefore the default value won't be used, and as a result there won't be any remote name on the page in snippets.

But if the case `null in db` happens, the "Value" itself is already wrong at many places. What would you expect to see in the "git clone command" for this case?

IMO we should fix such edge from root cause, but not add a `Def` method.

wxiaoguang commented on modules/setting/config.go:
```diff
@@ -49,6 +49,7 @@ func DefaultOpenWithEditorApps() OpenWithEditorAppsType {
 
 type RepositoryStruct struct {
 	OpenWithEditorApps *config.Value[OpenWithEditorAppsType]
+	SnippetRemoteName  *config.Value[string]
```
Is `Snippet` a proper name?

IMO it is not a "snippet", it is for "git guide for users"

wxiaoguang commented on options/locale/locale_en-US.ini:
```diff
@@ -3419,6 +3419,9 @@ config.disable_gravatar = Disable Gravatar
 config.enable_federated_avatar = Enable Federated Avatars
 config.open_with_editor_app_help = The "Open with" editors for the clone menu. If left empty, the default will be used. Expand to see the default.
 
+config.repository_snippets = Repository Snippets
```
Sorry I am not a native speaker ....

wxiaoguang commented:
> I think `Empty Repo Page` makes more sense on settings page:

But it is also for "templates/repo/issue/view_content/pull_merge_instruction.tmpl", it is a guide for end users to checkout the repo into their local clone. So I think `it is for "git guide for users"`, but no idea how to express in English.

ilya-nurullin commented:
Okay, I've removed `val.Def()` and changed translation to `Git Guide`.

<img width="1132" height="930" alt="image" src="https://github.com/user-attachments/assets/c9c7fa60-d0f9-463d-9d63-d3a6b16b7ca5" />

lunny commented on services/context/repo.go:
```diff
@@ -542,6 +542,7 @@ func RepoAssignment(ctx *Context) {
 	ctx.Data["CanWriteIssues"] = ctx.Repo.CanWrite(unit_model.TypeIssues)
 	ctx.Data["CanWritePulls"] = ctx.Repo.CanWrite(unit_model.TypePullRequests)
 	ctx.Data["CanWriteActions"] = ctx.Repo.CanWrite(unit_model.TypeActions)
+	ctx.Data["SnippetRemoteName"] = setting.Config().Repository.SnippetRemoteName.Value(ctx)
```
You can use `$.SystemConfig` directly if there is no special requirements.
```go
  ctx.Data["SystemConfig"] = setting.Config()
```
